### PR TITLE
Refactored code to work with IDs, unit tests

### PR DESCRIPTION
**Pull request by KamilFilip**

### CODE REVIEW

File: lib/cart.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
+require_relative "./warehouse"
+require_relative '../services/fetchProduct'
+require_relative '../services/fetchWarehouse'
+
```
Comment (Kacper3331) [inline lib/cart.rb]: Change single quote to double quote.

File: lib/main.rb [hunk 2]
```diff
@@ -1,51 +1,31 @@
...
+  p "============================================"
+  cart.add_to_cart(warehouse.storage,product1.id)
+  cart.add_to_cart(warehouse.storage,product1.id)
+  cart.add_to_cart(warehouse.storage,product2.id)
+  cart.remove_from_cart(warehouse.storage,product1.id)
+  p "AFTER"
+  p warehouse.storage
+  p "============================================"
+  p Shop::FetchProduct.new(cart.products).call(product1.id)
+  p Shop::FetchWarehouse.new(warehouse.storage).call(product1.id)
```
Comment (Kacper3331) [inline lib/main.rb]: Try to add some empty lines between the lines, to group them better, also add space after commas.

File: spec/unit/warehouse_spec.rb [hunk 3]
```diff
@@ -1,3 +1,34 @@
...
+    end
+
+    it "add given id and quantity to empty storage" do
+      quantity = 3
+      warehouse.add_to_storage(product1.id,quantity)
+      expect(fetch.call(product1.id)[:id]).to eql(product1.id)
+      expect(fetch.call(product1.id)[:quantity]).to eql(quantity)
+
+    end
+
```
Comment (Kacper3331) [inline spec/unit/warehouse_spec.rb]: Add empty space before the `context` or `describe` in this file and others, also your `describe` ends in the wrong place, you should move it to `line 33`.

File: lib/cart.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      @products.each do |x|
+        prod = catalog.find{|product| product.id == x[:id]}
+        sum +=prod.price*x[:quantity]
+      end
+      p "Total sum: #{sum}"
+      p "Total sum with VAT: #{sum_with_vat(sum)}"
+    end
+    def sum_with_vat(sum)
+      sum_vat = sum * 1.23
+    end
```
Comment (Kacper3331) [inline lib/cart.rb]: Check other files too, and add empty space between methods.

File: lib/cart.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+          else
+            @products << {id: id, quantity: 1}
+            selected_prod[:quantity]-=1
+          end
+        else
+          @products << {id: id, quantity: 1}
+          selected_prod[:quantity]-=1
+        end
+      else
+        no_item()
```
Comment (wafcio) [inline lib/cart.rb]: You shouldn't write `()` when method doesn't have arguments then you can write only method name.

File: lib/cart.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      end
+    end
+
+    def remove_from_cart(storage, id)
+      if(@products.any?)
+        prod = @products.find{|produce| produce[:id] == id}
+        selected_prod = storage.find{|selected| selected[:id] == id}
+        if prod[:quantity] >1
+          prod[:quantity]-=1
+          selected_prod[:quantity]+=1
```
Comment (marcindrozd) [inline lib/cart.rb]: Add spacing for better readability in the 3 above lines between the variables, signs and numbers. E.g. `selected_prod[:quantity] += 1`

File: lib/cart.rb [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        else
+          @products.delete(prod)
+        end
+      else
+        empty_cart_msg()
```
Comment (marcindrozd) [inline lib/cart.rb]: Same as mentioned above. Skip () for methods without arguments.

File: lib/cart.rb [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    end
+
+    def no_item
+      p "sorry we do not have that item in stock"
+    end
+
+    def empty_cart_msg
+      p "Your cart is empty"
+    end
+    def info(catalog)
```
Comment (marcindrozd) [inline lib/cart.rb]: Add empty line between methods.

File: services/fetchProduct.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+    def initialize(products)
+      @products = products
+    end
+
+    def call(id)
+      p @products
+     @products.find do |product|
+       product[:id] == id
+     end
```
Comment (marcindrozd) [inline services/fetchProduct.rb]: Check indentation here. Something is wrong.

File: services/fetchWarehouse.rb [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  class FetchWarehouse
+
+    def initialize(storage)
+      @storage = storage
+    end
+
+    def call(id)
+      @storage.find do |produce|
+       produce[:id] == id
+     end
```
Comment (marcindrozd) [inline services/fetchWarehouse.rb]: And indentation here too.

File: spec/unit/cart_spec.rb [hunk 11]
```diff
@@ -1,22 +1,24 @@
...
-  let(:product1){Product.new("Ball",12)}
-  let(:product2){Product.new("Glass",21)}
-  # it "should initialize empty @products array" do
-  #   expect (Cart.new.products).to eql([])
-  # end
+require_relative'../../lib/cart'
+
+RSpec.describe Shop::Cart do
+  let(:cart){Shop::Cart.new}
+  it "#should initialize empty array" do
```
Comment (marcindrozd) [inline spec/unit/cart_spec.rb]: We use `#` only in describes when we have method names like `describe "#add_to_basket"`.
Also `should` should be skipped in examples ;) just `it "initializes an empty array"`

File: spec/unit/cart_spec.rb [hunk 12]
```diff
@@ -1,22 +1,24 @@
...
+
     end
   end
+  context "#info" do
+  it "text" do
+
+  end
+  end
+
 end
```
Comment (marcindrozd) [inline spec/unit/cart_spec.rb]: Review and add empty lines between specs for easier readability.

File: spec/unit/cart_spec.rb [hunk 13]
```diff
@@ -1,22 +1,24 @@
...
-    #   }.to_not raise_error
-    # end
-    it "raises error for invalid number of args" do
-      expect{
-        cart.add_to_cart(nil)
-      }.to raise_error(ArgumentError)
+    it "text" do
+    end
+  end
+  context "#remove_from_cart" do
```
Comment (marcindrozd) [inline spec/unit/cart_spec.rb]: This should be `describe`. Context is better used for describing some condition, e.g.:

```ruby
describe "#remove_from_cart" do
  context "when cart is empty" do
  end

  context "when there is something in cart" do
  end
end
```

File: spec/unit/fetchWarehouse_spec.rb [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    warehouse.add_to_storage(product1.id,3)
+    expect(fetch.call(product1.id)[:id]).to eql(product1.id)
+  end
+  it "fetched product from warehouse quantity should return proper value" do
+    quantity = 3
+    warehouse.add_to_storage(product1.id,quantity)
+    expect(fetch.call(product1.id)[:quantity]).to eql(quantity)
+
+  end
+end
```
Comment (marcindrozd) [inline spec/unit/fetchWarehouse_spec.rb]: Again, please review spacing and lines between these. The one before `end` is not necessary while between `it` is important. Also add spaces after commas.

File: lib/cart.rb [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+module Shop
+  class Cart
+    attr_reader :products
+
+    def initialize
+      @products = []
+    end
+
+    def add_to_cart(storage, id)
+      selected_prod = storage.find{|selected| selected[:id] == id}
```
Comment (smt116) [inline lib/cart.rb]: missing spacebars between curly brackets

File: lib/cart.rb [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+      if selected_prod[:quantity] >=1
+        if @products.any?
+          prod = @products.find{|produce| produce[:id] == id}
+          if prod
+            prod[:quantity]+=1
```
Comment (smt116) [inline lib/cart.rb]: missing spacebars around `+=`

File: lib/cart.rb [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+            selected_prod[:quantity]-=1
+          else
+            @products << {id: id, quantity: 1}
+            selected_prod[:quantity]-=1
+          end
+        else
+          @products << {id: id, quantity: 1}
+          selected_prod[:quantity]-=1
+        end
```
Comment (smt116) [inline lib/cart.rb]: There are too many ifs and code is unreadable.

File: lib/warehouse.rb [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+        quantity: set_quantity(quantity)}
+    end
+
+    private
+    def set_quantity(quantity)
+      raise ArgumentError unless quantity.is_a?(Numeric)
+      raise ArgumentError unless quantity >= 0
+      quantity
+    end
+    def next_id
```
Comment (smt116) [inline lib/warehouse.rb]: missing empty line before `def`

**Pull request merged**